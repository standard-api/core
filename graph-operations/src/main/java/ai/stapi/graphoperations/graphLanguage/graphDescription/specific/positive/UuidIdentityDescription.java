package ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive;

import ai.stapi.graphoperations.graphLanguage.GraphBaseTypes;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescriptionParameters;
import java.util.ArrayList;
import java.util.List;

public class UuidIdentityDescription extends AbstractPositiveGraphDescription {

  public static final String SERIALIZATION_TYPE = "UuidIdentityDescription";
  private UuidIdentityDescriptionParameters parameters;

  public UuidIdentityDescription() {
    super(
        SERIALIZATION_TYPE,
        GraphBaseTypes.VALUE_TYPE,
        new ArrayList<>()
    );
    this.parameters = new UuidIdentityDescriptionParameters();
  }

  public UuidIdentityDescription(List<GraphDescription> children) {
    super(
        SERIALIZATION_TYPE,
        GraphBaseTypes.VALUE_TYPE,
        children
    );
    this.parameters = new UuidIdentityDescriptionParameters();
  }

  public UuidIdentityDescription(GraphDescription... children) {
    super(
        SERIALIZATION_TYPE,
        GraphBaseTypes.VALUE_TYPE,
        children
    );
    this.parameters = new UuidIdentityDescriptionParameters();
  }

  @Override
  public GraphDescriptionParameters getParameters() {
    return this.parameters;
  }
}
