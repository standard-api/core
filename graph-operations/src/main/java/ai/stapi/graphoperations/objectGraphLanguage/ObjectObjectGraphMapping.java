package ai.stapi.graphoperations.objectGraphLanguage;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import java.util.Map;

public class ObjectObjectGraphMapping extends AbstractObjectGraphMapping {

  public static final String SERIALIZATION_TYPE = "ObjectObjectGraphMapping";
  private Map<String, ObjectFieldDefinition> fields;

  private ObjectObjectGraphMapping() {

  }

  public ObjectObjectGraphMapping(
      GraphDescription graphDescription,
      Map<String, ObjectFieldDefinition> fields
  ) {
    super(
        graphDescription,
        SERIALIZATION_TYPE
    );
    this.fields = fields;
  }

  public ObjectObjectGraphMapping(
      GraphDescription graphDescription,
      Map<String, ObjectFieldDefinition> fields,
      String serializationType
  ) {
    super(
        graphDescription,
        serializationType
    );
    this.fields = fields;
  }

  public Map<String, ObjectFieldDefinition> getFields() {
    return fields;
  }

}
