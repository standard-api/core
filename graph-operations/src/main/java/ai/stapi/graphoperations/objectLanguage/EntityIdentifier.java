package ai.stapi.graphoperations.objectLanguage;

public class EntityIdentifier extends AbstractObjectDeclaration {

  public static final String SERIALIZATION_TYPE = "6a1a15dd-4aba-4c20-b883-2836d3e4cbd6";

  public EntityIdentifier() {
    super(SERIALIZATION_TYPE);
  }

  @Override
  public String getSerializationType() {
    return SERIALIZATION_TYPE;
  }

}
