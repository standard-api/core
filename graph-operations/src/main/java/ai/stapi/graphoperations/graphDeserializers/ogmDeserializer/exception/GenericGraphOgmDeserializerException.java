package ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exception;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.PositiveGraphDescription;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.graphoperations.ogmProviders.specific.objectGraphMappingMappingProviders.OgmGraphElementTypes;
import ai.stapi.graphoperations.graphReader.exception.GraphReaderException;
import ai.stapi.graphoperations.graphReader.readResults.ReadResult;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectObjectGraphMapping;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.graphoperations.serializationTypeProvider.specific.SpecificSerializationTypeProvider;
import java.util.List;
import org.apache.commons.lang3.StringUtils;

public class GenericGraphOgmDeserializerException extends RuntimeException {

  private GenericGraphOgmDeserializerException(String message) {
    super(message);
  }

  public static GenericGraphOgmDeserializerException becauseNodeTypeIsNotSupported(
      String nodeType) {
    return new GenericGraphOgmDeserializerException(
        "There are no supporting "
            + SpecificSerializationTypeProvider.class.getSimpleName()
            + " provider for node of type '"
            + nodeType + "'."
    );
  }

  public static GenericGraphOgmDeserializerException becauseNodeTypeIsSupportedByMultipleDeserializers(
      String nodeType,
      List<String> supportingDeserializerClassNames
  ) {
    return new GenericGraphOgmDeserializerException(
        "Node of type '" + nodeType
            + "' is supported by multiple deserializers: ["
            + StringUtils.join(supportingDeserializerClassNames, ", ")
            + "]."
    );
  }

  public static GenericGraphOgmDeserializerException becauseDeserializationHasToStartFromObjectGraphMapping(
      String providedNodeType) {
    return new GenericGraphOgmDeserializerException(
        "Deserialization has to start from "
            + ObjectObjectGraphMapping.class.getSimpleName()
            + " serialized under node '"
            + OgmGraphElementTypes.OGM_OBJECT_NODE
            + "'. Provided node type was: '"
            + providedNodeType + "'."
    );
  }

  public static GenericGraphOgmDeserializerException becauseThereIsMultipleGraphBranchesButMappingIsNotForList() {
    return new GenericGraphOgmDeserializerException(
        "Graph description led to multiple graph branches but provided "
            + ObjectGraphMapping.class.getSimpleName()
            + " did not point to list."
    );
  }

  public static GenericGraphOgmDeserializerException becauseGraphTraversingLedToMultipleBranches(
      String elementType) {
    return new GenericGraphOgmDeserializerException(
        GraphDescription.class.getSimpleName()
            + " inside " + ObjectObjectGraphMapping.class.getSimpleName()
            + " led from element of type '" + elementType
            + "' to multiple graph branches and that is not allowed."
    );
  }

  public static GenericGraphOgmDeserializerException becauseGraphReadingAccordingToOgmFailed(
      String currentElementType,
      PositiveGraphDescription declaration,
      GraphReaderException exception
  ) {
    return new GenericGraphOgmDeserializerException(
        "Traversing graph from element of type '" + currentElementType
            + "' using " + declaration.getClass().getSimpleName()
            + " failed. Ogm doesn't correspond to the actual graph structure saved. "
            + System.lineSeparator()
            + "Cause: " + exception.getMessage()
    );
  }

  public static GenericGraphOgmDeserializerException becauseElementWasNotFound(
      UniqueIdentifier graphElementId) {
    return new GenericGraphOgmDeserializerException(
        "Graph element with id [" + graphElementId
            + "] was not found in graph. This should have never happen!"
    );
  }

  public static GenericGraphOgmDeserializerException givenReadResultIsNotGraphElement() {
    return new GenericGraphOgmDeserializerException(
        ReadResult.class.getSimpleName() + " does not contain traversable graph element in it."
    );
  }

  public static GenericGraphOgmDeserializerException becauseGraphTraversingLedToNoBranches(
      String type) {
    return new GenericGraphOgmDeserializerException(
        GraphDescription.class.getSimpleName()
            + " inside " + ObjectObjectGraphMapping.class.getSimpleName()
            + " led from element of type '" + type
            + "' to no graph branches and that is not allowed."
    );
  }
}
