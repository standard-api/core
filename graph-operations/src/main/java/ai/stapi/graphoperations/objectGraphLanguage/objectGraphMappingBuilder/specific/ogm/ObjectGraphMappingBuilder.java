package ai.stapi.graphoperations.objectGraphLanguage.objectGraphMappingBuilder.specific.ogm;

import ai.stapi.graphoperations.graphLanguage.graphDescription.GraphDescription;
import ai.stapi.graphoperations.graphLanguage.graphDescription.graphDescriptionBuilder.GraphDescriptionBuilder;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.NullGraphDescription;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectFieldDefinition;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectObjectGraphMapping;
import ai.stapi.graphoperations.objectGraphLanguage.objectGraphMappingBuilder.GenericOGMBuilder;
import ai.stapi.graphoperations.objectGraphLanguage.objectGraphMappingBuilder.specific.SpecificObjectGraphMappingBuilder;
import java.util.HashMap;
import java.util.Map;

public class ObjectGraphMappingBuilder implements SpecificObjectGraphMappingBuilder {

  private final Map<String, ObjectFieldDefinitionBuilder> fields = new HashMap<>();
  private GraphDescription graphDescription;

  public ObjectFieldDefinitionBuilder addField(String fieldName) {
    var newField = new ObjectFieldDefinitionBuilder();
    this.fields.put(fieldName, newField);
    return newField;
  }

  public ObjectFieldDefinitionBuilder addField(String fieldName,
      ObjectFieldDefinitionBuilder fieldBuilder) {
    this.fields.put(fieldName, fieldBuilder);
    return fieldBuilder;
  }

  @Override
  public void setNewGraphDescription(GraphDescription graphDescription) {
    this.graphDescription = graphDescription;
  }

  public ObjectGraphMapping build() {
    if (this.graphDescription == null) {
      this.graphDescription = new NullGraphDescription();
    }
    var resolvedFields = new HashMap<String, ObjectFieldDefinition>();
    this.fields.forEach(
        (key, value) -> resolvedFields.put(key, value.build())
    );
    return new ObjectObjectGraphMapping(
        this.graphDescription,
        resolvedFields
    );
  }

  @Override
  public SpecificObjectGraphMappingBuilder getEmptyCopy() {
    return new ObjectGraphMappingBuilder();
  }

  @Override
  public boolean supports(ObjectGraphMapping objectGraphMapping) {
    return objectGraphMapping instanceof ObjectObjectGraphMapping;
  }

  @Override
  public SpecificObjectGraphMappingBuilder copyObjectGraphMappingAsBuilder(
      ObjectGraphMapping objectGraphMapping, GenericOGMBuilder genericBuilder) {
    var objectMapping = (ObjectObjectGraphMapping) objectGraphMapping;
    var builder = new ObjectGraphMappingBuilder();
    builder.setGraphDescription(objectGraphMapping.getGraphDescription());
    objectMapping.getFields().forEach(
        (key, value) -> {
          var copiedField = new ObjectFieldDefinitionBuilder();
          copiedField.setRelation(value.getRelation());
          copiedField.setChild(
              genericBuilder.copyGraphMappingAsBuilder(value.getFieldObjectGraphMapping()));
          builder.addField(key, copiedField);
        }
    );
    return builder;
  }

  @Override
  public GraphDescription getGraphDescription() {
    return this.graphDescription;
  }

  public ObjectGraphMappingBuilder setGraphDescription(GraphDescription graphDescription) {
    this.graphDescription = graphDescription;
    return this;
  }

  public ObjectGraphMappingBuilder setGraphDescription(GraphDescriptionBuilder builder) {
    this.graphDescription = builder.build();
    return this;
  }
}
