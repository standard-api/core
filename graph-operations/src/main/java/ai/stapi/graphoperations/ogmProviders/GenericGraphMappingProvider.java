package ai.stapi.graphoperations.ogmProviders;

import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.graphoperations.ogmProviders.exception.GraphMappingProviderException;
import ai.stapi.graphoperations.ogmProviders.specific.SpecificGraphMappingProvider;
import ai.stapi.graphoperations.ogmProviders.specific.dynamicObjectGraphMappingProvider.DynamicOgmProvider;
import ai.stapi.serialization.AbstractSerializableObject;
import ai.stapi.serialization.SerializableObject;
import java.util.List;
import java.util.Map;

public class GenericGraphMappingProvider {

  private final List<SpecificGraphMappingProvider> mappingProviders;

  public GenericGraphMappingProvider(
      List<SpecificGraphMappingProvider> mappingProviders
  ) {
    this.mappingProviders = mappingProviders;
  }

  public ObjectGraphMapping provideGraphMapping(Map<String, Object> serializableObjectMap) {
    if (!serializableObjectMap.containsKey(
        AbstractSerializableObject.NAME_OF_FIELD_WITH_SERIALIZATION_TYPE)) {
      throw GraphMappingProviderException.becauseProvidedObjectMapDoesNotContainTypeField();
    }
    var serializationType = (String) serializableObjectMap.get(
        AbstractSerializableObject.NAME_OF_FIELD_WITH_SERIALIZATION_TYPE);
    return this.provideGraphMapping(serializationType);
  }

  public ObjectGraphMapping provideGraphMapping(Map<String, Object> serializableObjectMap,
      String fieldName) {
    if (!serializableObjectMap.containsKey(
        AbstractSerializableObject.NAME_OF_FIELD_WITH_SERIALIZATION_TYPE)) {
      throw GraphMappingProviderException.becauseProvidedObjectMapDoesNotContainTypeField();
    }
    var serializationType = (String) serializableObjectMap.get(
        AbstractSerializableObject.NAME_OF_FIELD_WITH_SERIALIZATION_TYPE);
    return this.provideGraphMapping(serializationType, fieldName);
  }

  public ObjectGraphMapping provideGraphMapping(SerializableObject object) {
    return this.provideGraphMapping(object.getSerializationType());
  }

  public ObjectGraphMapping provideGraphMapping(SerializableObject object, String fieldName) {
    return this.provideGraphMapping(object.getSerializationType(), fieldName);
  }

  public ObjectGraphMapping provideGraphMapping(String serializationType, String fieldName) {
    var provider = this.getSupportingProvider(serializationType);
    return provider.provideGraphMapping(serializationType, fieldName);
  }

  public ObjectGraphMapping provideGraphMapping(String serializationType) {
    var provider = this.getSupportingProvider(serializationType);
    return provider.provideGraphMapping(serializationType);
  }

  public boolean supports(String serializationType) {
    return this.mappingProviders.stream()
        .anyMatch(provider -> provider.supports(serializationType));
  }

  private SpecificGraphMappingProvider getSupportingProvider(String serializationType) {
    var listOfSupportingProviders = this.mappingProviders.stream()
        .filter(specificObjectGraphMapper -> specificObjectGraphMapper.supports(serializationType))
        .toList();
    if (listOfSupportingProviders.isEmpty()) {
      throw GraphMappingProviderException.becauseThereIsNoSupportingSpecificGraphMappingProvider(
          serializationType);
    }
    var staticProviders = listOfSupportingProviders.stream()
        .filter(provider -> (!(provider instanceof DynamicOgmProvider)))
        .toList();
    if (staticProviders.size() > 1) {
      throw GraphMappingProviderException.becauseThereAreMoreThanOneSpecificGraphMappingProviders(
          serializationType,
          listOfSupportingProviders
      );
    }
    if (staticProviders.size() == 1) {
      return staticProviders.get(0);
    }
    var dynamicProvider = listOfSupportingProviders.stream()
        .filter(DynamicOgmProvider.class::isInstance)
        .findAny()
        .orElse(null);
    if (dynamicProvider != null) {
      return dynamicProvider;
    }
    throw GraphMappingProviderException.becauseThereIsNoSupportingSpecificGraphMappingProvider(
        serializationType
    );
  }

}
