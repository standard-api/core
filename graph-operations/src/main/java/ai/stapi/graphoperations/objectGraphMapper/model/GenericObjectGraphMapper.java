package ai.stapi.graphoperations.objectGraphMapper.model;

import ai.stapi.graphoperations.graphLanguage.graphDescription.graphDescriptionBuilder.GraphDescriptionBuilder;
import ai.stapi.graphoperations.graphWriter.GenericGraphWriter;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectGraphMapping;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectObjectGraphMapping;
import ai.stapi.graphoperations.objectGraphMapper.model.exceptions.ObjectGraphMapperException;
import ai.stapi.graphoperations.ogmProviders.GenericGraphMappingProvider;
import ai.stapi.serialization.SerializableObject;
import java.util.List;
import java.util.Map;

public class GenericObjectGraphMapper {

  private final List<SpecificObjectGraphMapper> specificObjectGraphMappers;
  private final GenericGraphWriter writer;
  private final GenericGraphMappingProvider mappingProvider;

  public GenericObjectGraphMapper(
      List<SpecificObjectGraphMapper> specificObjectGraphMappers,
      GenericGraphWriter writer,
      GenericGraphMappingProvider mappingProvider
  ) {
    this.specificObjectGraphMappers = specificObjectGraphMappers;
    this.writer = writer;
    this.mappingProvider = mappingProvider;
  }

  public GraphMappingResult mapToGraph(
      ObjectGraphMapping objectGraphMapping,
      Object object
  ) {
    return this.mapToGraph(objectGraphMapping, object, MissingFieldResolvingStrategy.STRICT);
  }

  public GraphMappingResult mapToGraph(
      ObjectGraphMapping objectGraphMapping,
      Object object,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    if (object == null) {
      throw ObjectGraphMapperException.becausegivenObjectIsNull(objectGraphMapping);
    }
    if (!(objectGraphMapping instanceof ObjectObjectGraphMapping)) {
      throw ObjectGraphMapperException.becauseFirstObjectGraphMappingIsNotObject(
          objectGraphMapping);
    }
    var builder = new GraphDescriptionBuilder();
    this.resolveInternally(
        objectGraphMapping,
        Map.entry("", object),
        builder,
        missingFieldResolvingStrategy
    );
    var graph = this.writer.createGraph(
        builder.getOnlyPositiveGraphDescriptions()
    );

    var elementsForRemoval = this.writer.createElementsForRemoval(
        builder.getOnlyRemovalGraphDescriptions()
    );
    return new GraphMappingResult(graph, elementsForRemoval);
  }

  public GraphMappingResult mapToGraph(SerializableObject object) {
    return this.mapToGraph(object, MissingFieldResolvingStrategy.STRICT);
  }

  public GraphMappingResult mapToGraph(
      SerializableObject object,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    var objectGraphMapping = this.mappingProvider.provideGraphMapping(object.getSerializationType());
    return this.mapToGraph(objectGraphMapping, object, missingFieldResolvingStrategy);
  }

  public GraphMappingResult mapToGraph(
      Object object,
      String serializationType,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    var objectGraphMapping = this.mappingProvider.provideGraphMapping(serializationType);
    return this.mapToGraph(objectGraphMapping, object, missingFieldResolvingStrategy);
  }

  public void resolveInternally(
      ObjectGraphMapping graphMapping,
      Map.Entry<String, Object> fieldEntry,
      GraphDescriptionBuilder builder,
      MissingFieldResolvingStrategy missingFieldResolvingStrategy
  ) {
    var supportingResolver = this.getSupportingSpecificMapper(graphMapping);
    supportingResolver.createGraphDescriptionWithValues(
        graphMapping,
        fieldEntry,
        builder,
        missingFieldResolvingStrategy
    );
  }

  private SpecificObjectGraphMapper getSupportingSpecificMapper(
      ObjectGraphMapping objectGraphMapping) {
    var supportingGraphWriters = this.specificObjectGraphMappers.stream()
        .filter(
            specificObjectGraphMapper -> specificObjectGraphMapper.supports(objectGraphMapping));

    var listOfSupportingResolvers = supportingGraphWriters.toList();
    if (listOfSupportingResolvers.size() == 0) {
      throw ObjectGraphMapperException.becauseNoSupportingSpecificResolverForGivenDeclaration(
          objectGraphMapping);
    }
    if (listOfSupportingResolvers.size() > 1) {
      throw ObjectGraphMapperException.becauseMoreThanOneSpecificResolverForGivenDeclaration(
          objectGraphMapping,
          listOfSupportingResolvers
      );
    }
    return listOfSupportingResolvers.get(0);
  }
}
