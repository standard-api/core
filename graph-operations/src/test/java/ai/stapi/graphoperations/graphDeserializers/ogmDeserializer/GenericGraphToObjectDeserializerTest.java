package ai.stapi.graphoperations.graphDeserializers.ogmDeserializer;

import ai.stapi.graphoperations.exampleImplementations.ExampleDto;
import ai.stapi.graphoperations.exampleImplementations.ExampleDtoWithListAttribute;
import ai.stapi.graphoperations.exampleImplementations.ExampleDtoWithListOfStrings;
import ai.stapi.graphoperations.exampleImplementations.ExampleNestedDto;
import ai.stapi.graphoperations.exampleImplementations.ExampleNestedDtoWithInterface;
import ai.stapi.graphoperations.exampleImplementations.ExampleNestedDtoWithListOfInterfaces;
import ai.stapi.graphoperations.exampleImplementations.ExampleNestedDtoWithMapOfInterfaces;
import ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exampleImplementations.exampleOgmProviders.ExampleDtoWithListAttributeSpecificGraphMappingProvider;
import ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exampleImplementations.exampleOgmProviders.ExampleNestedDtoSpecificGraphMappingProvider;
import ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exampleImplementations.exampleOgmProviders.ExampleNestedDtoWithInterfaceSpecificGraphMappingProvider;
import ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exampleImplementations.exampleOgmProviders.ExampleNestedDtoWithListOfInterfacesGraphMappingProvider;
import ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exampleImplementations.exampleOgmProviders.ExampleNestedDtoWithMapOfInterfacesGraphMappingProvider;
import ai.stapi.graphoperations.graphLanguage.graphDescription.graphDescriptionBuilder.GraphDescriptionBuilder;
import ai.stapi.graphoperations.graphLanguage.graphDescription.specific.positive.UuidIdentityDescription;
import ai.stapi.graph.Graph;
import ai.stapi.graph.inMemoryGraph.InMemoryGraphRepository;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.graph.traversableGraphElements.TraversableNode;
import ai.stapi.graphoperations.graphDeserializers.ogmDeserializer.exception.GenericGraphOgmDeserializerException;
import ai.stapi.graphoperations.objectGraphLanguage.ObjectObjectGraphMapping;
import ai.stapi.graphoperations.objectGraphLanguage.objectGraphMappingBuilder.specific.ogm.ObjectGraphMappingBuilder;
import ai.stapi.graphoperations.objectGraphMapper.model.GenericObjectGraphMapper;
import ai.stapi.graphoperations.objectLanguage.EntityIdentifier;
import ai.stapi.objectRenderer.infrastructure.objectToJsonStringRenderer.ObjectToJSonStringOptions;
import ai.stapi.graphoperations.ogmProviders.GenericGraphMappingProvider;
import ai.stapi.graphoperations.ogmProviders.specific.objectGraphMappingMappingProviders.OgmGraphElementTypes;
import ai.stapi.serialization.AbstractSerializableObject;
import ai.stapi.test.integration.IntegrationTestCase;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;
import org.springframework.beans.factory.annotation.Autowired;

class GenericGraphToObjectDeserializerTest extends IntegrationTestCase {

  @Autowired
  private GenericGraphToObjectDeserializer deserializer;

  @Autowired
  private GenericObjectGraphMapper objectGraphMapper;

  @Autowired
  private GenericGraphMappingProvider mappingProvider;

  @Test
  void itThrowsErrorWhenNodeTypeIsNotSupportedForDeserialization() {
    var graph = new Graph(new Node("random_type")).traversable();
    Executable throwable = () -> deserializer.deserialize(
        graph.loadAllNodes("random_type").get(0),
        AbstractSerializableObject.class,
        graph
    );
    this.thenExceptionMessageApproved(GenericGraphOgmDeserializerException.class, throwable);
  }

  @Test
  void itCanDeserializeSimpleDto() {
    var graph = this.objectGraphMapper.mapToGraph(
        new ExampleDto("Ivan Mladek")
    ).getGraph().traversable();
    var result = this.deserializer.deserialize(
        graph.loadAllNodes("example_dto_node").get(0),
        ExampleDto.class,
        graph
    );
    this.thenObjectApproved(result, this.getThisOptions());
  }

  @Test
  void itCanDeserializeDtoWithNestedDto() {
    var graph = this.objectGraphMapper.mapToGraph(
        new ExampleNestedDto(
            "Master name",
            new ExampleDto("Ivan Mladek")
        )
    ).getGraph().traversable();
    var result = this.deserializer.deserialize(
        graph.loadAllNodes(ExampleNestedDtoSpecificGraphMappingProvider.NODE_TYPE).get(0),
        ExampleNestedDto.class,
        graph
    );
    this.thenObjectApproved(result, this.getThisOptions());
  }

  @Test
  void itCanDeserializeDtoWithListAttributes() {
    var graph = this.objectGraphMapper.mapToGraph(
        new ExampleDtoWithListAttribute(List.of(
            "Tag 1 ", "Tag 2", "Tag 3"
        )
        )
    ).getGraph().traversable();
    var result = this.deserializer.deserialize(
        graph.loadAllNodes(ExampleDtoWithListAttributeSpecificGraphMappingProvider.NODE_TYPE).get(0),
        ExampleDtoWithListAttribute.class,
        graph
    );
    this.thenObjectApproved(result, this.getThisOptions());
  }

  @Test
  void itCanDeserializeDtoWithInterface() {
    var testedDto = new ExampleNestedDtoWithInterface(
        "Interface dto",
        new ExampleDtoWithListOfStrings(
            List.of("Marek", "Zdenka", "Petr")
        )
    );
    var graph = this.objectGraphMapper.mapToGraph(testedDto).getGraph().traversable();
    var result = this.deserializer.deserialize(
        graph.loadAllNodes(ExampleNestedDtoWithInterfaceSpecificGraphMappingProvider.NODE_TYPE)
            .get(0),
        ExampleNestedDtoWithInterface.class,
        graph
    );
    var nestedDto = (ExampleDtoWithListOfStrings) result.getChildDto();
    Assertions.assertEquals(3, nestedDto.getNames().size());
    Assertions.assertEquals(ExampleDtoWithListOfStrings.SERIALIZATION_TYPE,
        nestedDto.getSerializationType());
    Assertions.assertEquals(ExampleNestedDtoWithInterface.SERIALIZATION_TYPE,
        result.getSerializationType());
  }

  @Test
  void itCanDeserializeDtoWithListOfInterfaces() {
    var nestedDto1 = new ExampleDto("Pushkin");
    var nestedDto2 = new ExampleDto("Metrostav");
    var nestedDto3 = new ExampleDto("CocaCola");
    var testedDto = new ExampleNestedDtoWithListOfInterfaces(
        "dto with list of interfaces",
        List.of(nestedDto1, nestedDto2, nestedDto3)
    );
    var graph = this.objectGraphMapper.mapToGraph(testedDto).getGraph().traversable();
    var deserializedDto = this.deserializer.deserialize(
        graph.loadAllNodes(ExampleNestedDtoWithListOfInterfacesGraphMappingProvider.NODE_TYPE)
            .get(0),
        ExampleNestedDtoWithListOfInterfaces.class,
        graph
    );
    Assertions.assertEquals(3, deserializedDto.getChildDtos().size());
    Assertions.assertEquals(ExampleNestedDtoWithListOfInterfaces.SERIALIZATION_TYPE,
        deserializedDto.getSerializationType());
  }

  @Test
  void itCanDeserializeDtoWithMapOfInterfaces() {
    var nestedDto1 = new ExampleDto("Pushkin");
    var nestedDto2 = new ExampleDto("Metrostav");
    var nestedDto3 = new ExampleDto("CocaCola");
    var testedDto = new ExampleNestedDtoWithMapOfInterfaces(
        "dto with list of interfaces",
        Map.ofEntries(
            Map.entry(nestedDto1.getName(), nestedDto1),
            Map.entry(nestedDto2.getName(), nestedDto2),
            Map.entry(nestedDto3.getName(), nestedDto3)
        )
    );
    var graph = this.objectGraphMapper.mapToGraph(testedDto).getGraph().traversable();
    var deserializedDto = this.deserializer.deserialize(
        graph.loadAllNodes(ExampleNestedDtoWithMapOfInterfacesGraphMappingProvider.NODE_TYPE)
            .get(0),
        ExampleNestedDtoWithMapOfInterfaces.class,
        graph
    );
    Assertions.assertEquals(3, deserializedDto.getChildDtos().size());
    Assertions.assertEquals(ExampleNestedDtoWithMapOfInterfaces.SERIALIZATION_TYPE,
        deserializedDto.getSerializationType());
    Assertions.assertEquals(ExampleDto.SERIALIZATION_TYPE,
        deserializedDto.getChildDtos().get(nestedDto1.getName()).getSerializationType());
  }

  @Test
  void itCanDeserializeSimpleOgm() {
    var graph = this.getSavedOgmSimpleExample();
    var rootNode = this.getOgmRootNode(graph);
    var result = this.deserializer.deserialize(
        rootNode,
        ObjectObjectGraphMapping.class,
        graph
    );

    this.thenObjectApproved(result, this.getThisOptions());
  }

  @Test
  void itCanDeserializeComplexOgm() {
    var graph = this.getSavedOgmMoreComplexExample();
    var rootNode = this.getOgmRootNode(graph);
    var result = this.deserializer.deserialize(
        rootNode,
        ObjectObjectGraphMapping.class,
        graph
    );
    this.thenObjectApproved(result);
  }

  @Test
  void itCanDeserializeComplexOgmWithIdentityIdentifier() {
    var graph = this.getSavedOgmMoreComplexWithEntityIdentifierExample();
    var rootNode = this.getOgmRootNode(graph);
    var result = this.deserializer.deserialize(
        rootNode,
        ObjectObjectGraphMapping.class,
        graph
    );
    this.thenObjectApproved(result, this.getThisOptions());
  }

  private TraversableNode getOgmRootNode(InMemoryGraphRepository contextualGraph) {
    var rootNode = contextualGraph.loadAllNodes(OgmGraphElementTypes.OGM_OBJECT_NODE).stream()
        .filter(node -> node.getEdges().stream()
            .noneMatch(edge -> edge.getNodeToId().equals(node.getId())))
        .toList();
    return rootNode.get(0);
  }

  private InMemoryGraphRepository getSavedOgmSimpleExample() {
    var builder = new ObjectGraphMappingBuilder()
        .setGraphDescription(new GraphDescriptionBuilder().addNodeDescription("example_node")
        );
    var mapping =
        this.mappingProvider.provideGraphMapping(ObjectObjectGraphMapping.SERIALIZATION_TYPE);
    return this.objectGraphMapper.mapToGraph(
        mapping,
        builder.build()
    ).getGraph().traversable();
  }

  private InMemoryGraphRepository getSavedOgmMoreComplexExample() {
    var builder = new ObjectGraphMappingBuilder()
        .setGraphDescription(new GraphDescriptionBuilder().addNodeDescription("example_node"));
    builder.addField("randomFieldName")
        .addObjectAsObjectFieldMapping()
        .addField("nestedObjectField")
        .addLeafAsObjectFieldMapping()
        .setGraphDescription(
            new GraphDescriptionBuilder()
                .addLeafAttribute("nestedAttribute")
                .addStringAttributeValue()
        );
    var mapping =
        this.mappingProvider.provideGraphMapping(ObjectObjectGraphMapping.SERIALIZATION_TYPE);
    return this.objectGraphMapper.mapToGraph(
        mapping,
        builder.build()
    ).getGraph().traversable();
  }

  private InMemoryGraphRepository getSavedOgmMoreComplexWithEntityIdentifierExample() {
    var builder = new ObjectGraphMappingBuilder();
    builder.addField("entityIdentifier")
        .setRelation(new EntityIdentifier())
        .addObjectAsObjectFieldMapping()
        .setGraphDescription(new GraphDescriptionBuilder().addNodeDescription("dto_identity"))
        .addField("id")
        .addLeafAsObjectFieldMapping()
        .setGraphDescription(new UuidIdentityDescription());
    var mapping =
        this.mappingProvider.provideGraphMapping(ObjectObjectGraphMapping.SERIALIZATION_TYPE);
    return this.objectGraphMapper.mapToGraph(
        mapping,
        builder.build()
    ).getGraph().traversable();
  }

  private ObjectToJSonStringOptions getThisOptions() {
    return new ObjectToJSonStringOptions(
        ObjectToJSonStringOptions.RenderFeature.HIDE_CREATED_AT,
        ObjectToJSonStringOptions.RenderFeature.SORT_FIELDS
    );
  }


}