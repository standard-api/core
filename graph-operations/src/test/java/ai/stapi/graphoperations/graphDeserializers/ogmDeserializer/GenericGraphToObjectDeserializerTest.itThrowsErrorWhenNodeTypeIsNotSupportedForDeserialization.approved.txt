There are no supporting SpecificSerializationTypeProvider provider for node of type 'random_type'.